src/messageschannel.cpp

```diff
@@ -185,3 +202,47 @@ void MatrixMessagesChannel::fetchHistory()
         }
     }
 }
+
+void MatrixMessagesChannel::onTypingChanged()
+{
+    if (m_room->usersTyping().isEmpty())
+    {
+        for(auto user: m_room->users())
+        {
+            const uint handle = m_connection->ensureContactHandle(user->id());
+            m_chatStateIface->chatStateChanged(handle, Tp::ChannelChatStateActive);
+        }
+        return;
+    }
+    for(auto user: m_room->usersTyping())
+    {
+        const uint handle = m_connection->ensureContactHandle(user->id());
+        m_chatStateIface->chatStateChanged(handle, Tp::ChannelChatStateComposing);
+    }
+}
+
+void MatrixMessagesChannel::reactivateLocalTyping()
+{
+    sendTypingNotification(true);
+}
+
+void MatrixMessagesChannel::setChatStateCallback(uint state, Tp::DBusError *error)
```

Comment:
I think this `Callback` word actually confuses a reader.
What I would expect from a "set chat state callback" method? I would expect a setter for some 'chat state' callback.
On the other hand this method actually set chat state (in Matrix backend). IMO `setChatState` sounds good as it is and it actually do what the name means. It is a coincide that the method can be used as a callback for the interface.